File: 프로그래머스/2/42626. 더 맵게/더 맵게.js
var index = 0;
var index2 = 0;

function min(scoville, new_scoville) {
    if((scoville[index] ?? Infinity) <= (new_scoville[index2] ?? Infinity)) {
        return scoville[index++];
    } else {
        return new_scoville[index2++];
    }
}

function get_min_scovilles(scoville, new_scoville) {
    return [min(scoville, new_scoville), min(scoville, new_scoville)];
}


function solution(scoville, K) {
    var answer = 0;
    let new_scoville = [];
    let scoville_length = scoville.length;
    
    
    scoville.sort((a, b) => a - b);
    
    if(scoville[0] >= K) {
        return 0;
    }
        
    while(1) {
        if(((scoville_length - index) + (answer - index2)) < 2) {
            break;
        }
        
        if((scoville[index] ?? Infinity) >= K && (new_scoville[index2] ?? Infinity) >= K) {
            break;
        }
        
        answer++;

        let [min, min2] = get_min_scovilles(scoville, new_scoville);
        new_scoville.push(min2 * 2 + min);
        
        // console.log(min, min2, index, index2);
        // console.log(scoville);
        // console.log(new_scoville);
        // console.log();
    }
    
    if(scoville[index] < K || new_scoville[index2] < K) {
        return -1;
    }
    
    return answer;
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.py
import heapq

def solution(scoville, K):
    s = []
    answer = 0
    
    for sco in scoville:
        heapq.heappush(s, sco)
    
    while len(s) > 1:
        lv1 = heapq.heappop(s)
        lv2 = heapq.heappop(s)
        
        if lv1 >= K:
            return answer
        
        answer += 1
        heapq.heappush(s, lv1 + (lv2 * 2))
        
    if not s or s[0] < K:
        return -1
    return answer